src/test/java/software/amazon/cloudwatchlogs/emf/logger/MetricsLoggerTest.java

```diff
@@ -88,101 +134,243 @@ public void testOverrideDefaultDimensions() {
 
-        Assert.assertEquals(1, sink.getContext().getDimensions().size());
-        Assert.assertEquals(1, sink.getContext().getDimensions().get(0).getDimensionKeys().size());
-        Assert.assertEquals(
+        assertEquals(1, sink.getContext().getDimensions().size());
+        assertEquals(1, sink.getContext().getDimensions().get(0).getDimensionKeys().size());
+        assertEquals(
                 dimensionValue,
                 sink.getContext().getDimensions().get(0).getDimensionValue(dimensionName));
     }
 
     @Test
-    public void testSetDimensionsAndPreserveDefault() {
+    void setDimensions_overridesPreviousDimensionsAndPreservesDefault()
+            throws InvalidDimensionException {
         String dimensionName = "dim";
         String dimensionValue = "dimValue";
         logger.putDimensions(DimensionSet.of("foo", "bar"));
         logger.setDimensions(true, DimensionSet.of(dimensionName, dimensionValue));
         logger.flush();
 
-        Assert.assertEquals(sink.getContext().getDimensions().size(), 1);
-        Assert.assertEquals(sink.getContext().getDimensions().get(0).getDimensionKeys().size(), 4);
-        Assert.assertEquals(
+        assertEquals(1, sink.getContext().getDimensions().size());
+        assertEquals(4, sink.getContext().getDimensions().get(0).getDimensionKeys().size());
+        assertEquals(
                 sink.getContext().getDimensions().get(0).getDimensionValue(dimensionName),
                 dimensionValue);
     }
 
     @Test
-    public void testSetNamespace() {
+    void setDimensions_clearsDefaultDimensions() throws InvalidMetricException {
+        MetricsLogger logger = new MetricsLogger(envProvider);
+        logger.setDimensions();
+        logger.putMetric("Count", 1);
+        logger.flush();
+        List<DimensionSet> dimensions = sink.getContext().getDimensions();
+
+        assertEquals(0, dimensions.size());
+        assertEquals(1, sink.getLogEvents().size());
+
+        String logEvent = sink.getLogEvents().get(0);
+        assertTrue(logEvent.contains("\"Dimensions\":[]"));
+    }
+
+    @Test
+    void flush_PreservesDimensions() throws InvalidDimensionException {
+        MetricsLogger logger = new MetricsLogger(envProvider);
+        logger.setDimensions(DimensionSet.of("Name", "Test"));
+        logger.flush();
+        expectDimension("Name", "Test");
+
+        logger.flush();
+        expectDimension("Name", "Test");
+    }
+
+    @Test
+    void flush_doesNotPreserveDimensions() throws InvalidDimensionException {
+        logger.putDimensions(DimensionSet.of("Name", "Test"));
+        logger.setFlushPreserveDimensions(false);
+
+        logger.flush();
+        assertEquals(4, sink.getContext().getDimensions().get(0).getDimensionKeys().size());
+        expectDimension("Name", "Test");
+
+        logger.flush();
+        assertEquals(3, sink.getContext().getDimensions().get(0).getDimensionKeys().size());
+        expectDimension("Name", null);
+    }
```

Comment:
I'm a bit confused by what is happening here, what is the size before the first flush on line 238?

Reply:
There are 4 dimensions before the flush, 1 custom dimension and 3 default dimensions.
When flushed, all 4 dimensions are _first_ sent and the context is then copied. Since we invoked `logger.setFlushPreserveDimensions(false)`, the newly created context will not include the 1 custom dimensions.
Therefore after the next flush we are only left with the 3 default dimensions.